### PR TITLE
Add GET API for allowable operations

### LINKED ISSUE #2455: API to return what operations user can perform such that UI enable/disable controls
UI disables the edit functionality of an entity by checking if the logged-in user 
is a owner of the entity or not. This is breaking Datasteward role as the user may have edit description operation allowed
but UI will show disabled button as the logged-in user is not a owner. 
We should provide an API which returns what operations for a given entity user can perform 
and not use the if the logged-in user is a owner or not

### PR DESCRIPTION
**Pull request by mithmatt**
- [x] Amend PolicyEvaluator to support getAllowedOperations()
- [x] Set up REST API for v1/config/permissions
- [x] Integrate the above with DefaultAuthorizer
- [x] Add unit tests

### Describe your changes :

I worked on this to fix #2455

image

#
### Type of change :

- [ ] Bug fix
- [x] Improvement
- [ ] New feature
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation

#
### Frontend Preview (Screenshots) :
<p align="center">For frontend related change, please link screenshots of your changes preview! Optional for backend related changes.
</p>

#
### Checklist:

- [x] I have read the [**CONTRIBUTING**](https://docs.open-metadata.org/open-source-community/developer) document.
- [x] I have performed a self-review of my own. 
- [x] I have tagged my reviewers below.
- [x] I have commented on my code, particularly in hard-to-understand areas.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] All new and existing tests passed.

#
### Reviewers

 Backend: @sureshms @harshach @Sachin-chaurasiya

### CODE REVIEW

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/security/Permission.java [hunk 1]
```diff
@@ -1,35 +0,0 @@
-/*
```
Comment (mithmatt) [inline catalog-rest-service/src/main/java/org/openmetadata/catalog/security/Permission.java:1]: Remove dead code

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/security/ResourceType.java [hunk 2]
```diff
@@ -1,30 +0,0 @@
-/*
```
Comment (mithmatt) [inline catalog-rest-service/src/main/java/org/openmetadata/catalog/security/ResourceType.java:1]: Remove dead code

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/resources/config/ConfigResource.java [hunk 3]
```diff
@@ -77,12 +82,28 @@ public AuthenticationConfiguration getAuthConfig(@Context UriInfo uriInfo, @Cont
...
+  public AuthorizerConfiguration getAuthorizerConfig() {
     AuthorizerConfiguration authorizerConfiguration = new AuthorizerConfiguration();
     if (catalogApplicationConfig.getAuthorizerConfiguration() != null) {
       authorizerConfiguration = catalogApplicationConfig.getAuthorizerConfiguration();
     }
     return authorizerConfiguration;
   }
+
+  @GET
+  @Path(("/permissions"))
```
Comment (mithmatt) [inline catalog-rest-service/src/main/java/org/openmetadata/catalog/resources/config/ConfigResource.java:94]: The new API

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/security/DefaultAuthorizer.java [hunk 4]
```diff
@@ -155,6 +158,26 @@ public boolean hasPermissions(
...
+  public List<MetadataOperation> listPermissions(AuthenticationContext ctx, EntityReference entityReference) {
+    validateAuthenticationContext(ctx);
+    try {
+      User user = getUserFromAuthenticationContext(ctx);
+      if (entityReference == null) {
+        return policyEvaluator.getAllowedOperations(user, null);
+      }
+      Object entity = Entity.getEntity(entityReference, new EntityUtil.Fields(List.of("tags", "owner")));
+      EntityReference owner = Entity.getEntityInterface(entity).getOwner();
+      if (owner == null || isOwnedByUser(user, owner)) {
```
Comment (mithmatt) [inline catalog-rest-service/src/main/java/org/openmetadata/catalog/security/DefaultAuthorizer.java:171]: If owner is null or if entity is owned by the user, allow all operations

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/security/NoopAuthorizer.java [hunk 5]
```diff
@@ -37,6 +38,12 @@ public boolean hasPermissions(
     return true;
   }
 
+  @Override
+  public List<MetadataOperation> listPermissions(AuthenticationContext ctx, EntityReference entityReference) {
+    // Return all operations.
+    return Stream.of(MetadataOperation.values()).collect(Collectors.toList());
```
Comment (mithmatt) [inline catalog-rest-service/src/main/java/org/openmetadata/catalog/security/NoopAuthorizer.java:44]: For NoopAuthorizer allow all

**Comment by mithmatt** [comment]
Unit test work in progress

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/security/DefaultAuthorizer.java [hunk 6]
```diff
@@ -155,6 +158,26 @@ public boolean hasPermissions(
     }
   }
 
+  @Override
+  public List<MetadataOperation> listPermissions(AuthenticationContext ctx, EntityReference entityReference) {
+    validateAuthenticationContext(ctx);
+    try {
+      User user = getUserFromAuthenticationContext(ctx);
+      if (entityReference == null) {
```
Comment (harshach) [inline catalog-rest-service/src/main/java/org/openmetadata/catalog/security/DefaultAuthorizer.java:166]: if user is admin or bot, short-circuit here and return all operations

**Comment by mithmatt** [comment]
@pmbrull  py tests are failing on this PR arbitrarily

**Comment by mithmatt** [comment]
Flaky tests from webhook

```
2022-01-30T03:08:50.4716147Z [INFO] 
2022-01-30T03:08:50.4780639Z [INFO] Results:
2022-01-30T03:08:50.4780890Z [INFO] 
2022-01-30T03:08:50.4781112Z [ERROR] Failures: 
2022-01-30T03:08:50.4783466Z [ERROR]   WebhookResourceTest.put_updateEndpointURL:178->EntityResourceTest.deleteEntity:1297->EntityResourceTest.deleteEntity:1307 Expected org.apache.http.client.HttpResponseException to be thrown, but nothing was thrown.
2022-01-30T03:08:50.4784114Z [INFO] 
2022-01-30T03:08:50.4784378Z [ERROR] Tests run: 774, Failures: 1, Errors: 0, Skipped: 0
2022-01-30T03:08:50.4784654Z [INFO] 
```

File: catalog-rest-service/src/test/java/org/openmetadata/catalog/resources/config/ConfigResourceTest.java [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+    Role dataConsumerRole =
+        roleResourceTest.getEntityByName(DATA_CONSUMER_ROLE_NAME, RoleResource.FIELDS, ADMIN_AUTH_HEADERS);
+    userResourceTest.createEntity(
+        userResourceTest
+            .createRequest(DATA_CONSUMER_USER_NAME, "", "", null)
+            .withRoles(List.of(dataConsumerRole.getId())),
+        ADMIN_AUTH_HEADERS);
+  }
+
+  @ParameterizedTest
```
Comment (mithmatt) [inline catalog-rest-service/src/test/java/org/openmetadata/catalog/resources/config/ConfigResourceTest.java:70]: image

**Comment by pmbrull** [comment]
Hi @mithmatt, taking a look

**Comment by pmbrull** [comment]
Hi @mithmatt, sorry for the issue. This https://github.com/open-metadata/OpenMetadata/pull/2499 should fix it

**Review by harshach — APPROVED** [review]

**Pull request merged**